### PR TITLE
Private journals and happiness API improvements

### PR DESCRIPTION
**Pull request by jonathanjma**
- Add endpoints for editing and deleting journal entries
- Add ability to recover journal entries when password is reset (either by setting up a recovery phrase or using the old password key)
- Improved happiness api endpoints (added format checks, edit/delete by date)
- Added/updated tests

### CODE REVIEW

File: happiness-backend/api/dao/journal_dao.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
 from api.models import Journal
 
+def get_journal_by_id(entry_id):
+    """
+    Returns a Journal entry object by ID.
+    """
+    return Journal.query.filter(Journal.id == entry_id).first()
```
Comment (zachseidner1) [inline happiness-backend/api/dao/journal_dao.py]: Very small thing, but generally it is advised to use `.filter_by` for simpler queries like this one.  [Source](https://stackoverflow.com/a/2128558/16776130)

File: happiness-backend/api/journal.py [hunk 2]
```diff
@@ -50,4 +51,42 @@ def get_entries(req, headers):
     page, count = req.get("page", 1), req.get("count", 10)
     # add password key to schema context so entries can be decrypted
     DecryptedJournalSchema.context['password_key'] = headers.get('password_key')
-    return journal_dao.get_entries_by_count(user.id, page, count)
\ No newline at end of file
+    return journal_dao.get_entries_by_count(user.id, page, count)
+
+@journal.put('/<int:id>')
```
Comment (zachseidner1) [inline happiness-backend/api/journal.py]: Why did you change the route for editing happiness to not include the ID, but the route for editing journals does? We should have consistency in the routes throughout our models.

File: happiness-backend/api/user.py [hunk 3]
```diff
@@ -213,18 +218,34 @@ def change_user_info(req, headers):
         except Exception as e:
             print(e)
             return failure_response('Invalid password key.', 400)
+    elif data_type == "key_recovery_phrase":
+        # Adds/Changes key recovery phrase to prevent loss of encrypted data on password reset.
+        try:
+            current_user.add_key_recovery(req.get("data"), headers.get("password_key"))
```
Comment (zachseidner1) [inline happiness-backend/api/user.py:224]: I think we should add all key recovery phrases as either all lowercase or all uppercase. This way if a user answers a security question in a different case it doesn't prevent them from accessing their data.

File: happiness-backend/tests/test_journals.py [hunk 4]
```diff
@@ -36,27 +37,41 @@ def auth_key_header(token, key):
 def test_e2e(init_client):
```
Comment (zachseidner1) [inline happiness-backend/tests/test_journals.py]: Is it possible to separate this out into more tests? This will make the code more readable and make it easier to skip tests when necessary.

File: happiness-backend/tests/test_journals.py [hunk 5]
```diff
@@ -92,4 +113,58 @@ def test_journals(init_client):
...
+    assert reset_password.status_code == 204
+
+    bad_get = client.get('/api/journal/', headers=auth_key_header(token, pwd_key))
+    assert bad_get.status_code == 400
+
+    get = client.get('/api/journal/', headers=auth_key_header(token, user.derive_pwd_key('W password')))
+    assert get.status_code == 200
+    assert get.json[1]['data'] == 'secret' and get.json[0]['data'] == 'secret2'
+
+def test_edit_delete(init_client):
```
Comment (zachseidner1) [inline happiness-backend/tests/test_journals.py]: I think these should be separated into `test_edit` and `test_delete` just to improve readability and have the code as loosely coupled as possible.

**Review by zachseidner1 — COMMENTED** [review]
Great changes, thanks so much for making the recovery phrase and improving code quality all around! I've commented about some things you may want to change, but other than that awesome job.

**Review by zachseidner1 — CHANGES_REQUESTED** [review]
Looks like happiness editing and deletion are broken on the deployed Vercel preview! Make sure that you test using the Vercel preview before submitting a PR.

**Comment by jonathanjma** [comment]
It's broken bc the prod backend isn't updated since it on my branch

**Comment by zachseidner1** [comment]
Wdym? The only migration you're performing on the prod db is for recovery options, which are separate from Happiness entries. I don't see why happiness entries are broken.

**Comment by jonathanjma** [comment]
No like a changed the api but the preview only uses the prod version of the api

**Comment by zachseidner1** [comment]
Oh right that makes sense. Ok sounds good, then just address some of my other comments and then LGTM!

File: happiness-backend/tests/test_journals.py (hunk 5 above)
Comment (jonathanjma) [reply happiness-backend/tests/test_journals.py]: you should have @Alex967CW fix his too then....

**Comment by zachseidner1** [comment]
Looks great! Thanks so much for addressing the PR comments. Great work and I really appreciate your improvements to the API overall.

**Review by zachseidner1 — APPROVED** [review]
W
